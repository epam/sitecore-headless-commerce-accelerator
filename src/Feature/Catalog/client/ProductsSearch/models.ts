//    Copyright 2020 EPAM Systems, Inc.
// 
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import { RouterProps } from 'react-router';

import * as Jss from 'Foundation/ReactJss/client';

import * as ProductsSearch from 'Feature/Catalog/client/Integration/ProductsSearch';

import * as commonModels from 'Feature/Catalog/client/Integration/common/models';

export interface ProductsSearchOwnProps extends RouterProps {}
export interface ProductsSearchStateProps {
  productSearchParams: ProductsSearch.Params;
  isLoading: boolean;
}
export interface ProductsSearchDispatchProps {
  ChangeRoute: (newRoute: string) => void;
}

export interface ProductsSearchProps
extends ProductsSearchOwnProps,
ProductsSearchStateProps,
ProductsSearchDispatchProps {}

export interface ProductsSearchOwnState extends Jss.SafePureComponentState {
  initialKeyword: string;
  keyword: string;
  submitted: boolean;
}

export interface AppState extends ProductsSearch.GlobalProductSearchState, commonModels.AppState {}
