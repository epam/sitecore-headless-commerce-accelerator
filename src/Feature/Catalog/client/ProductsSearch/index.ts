//    Copyright 2020 EPAM Systems, Inc.
// 
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import { withExperienceEditorChromes } from '@sitecore-jss/sitecore-jss-react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { compose } from 'recompose';
import { bindActionCreators } from 'redux';

import { ChangeRoute } from 'Foundation/ReactJss/client/SitecoreContext';

import * as ProductSearch from 'Feature/Catalog/client/Integration/ProductsSearch';

import ProductsSearchComponent from './Component';
import { AppState, ProductsSearchDispatchProps, ProductsSearchOwnProps, ProductsSearchStateProps } from './models';

import { LoadingStatus } from 'Foundation/Integration/client';

const mapStateToProps = (state: AppState, ownProps: ProductsSearchOwnProps) => {
  const productSearchParams = ProductSearch.productSearchParams(state);
  const status = ProductSearch.productsSearchStatus(state);
  const isLoading = status === LoadingStatus.Loading;

  return {
    isLoading,
    productSearchParams
  };
};

const mapDispatchToProps = (dispatch: any) =>
  bindActionCreators(
    {
      ChangeRoute,
    },
    dispatch
  );

const connectedToStore = connect<ProductsSearchStateProps, ProductsSearchDispatchProps, ProductsSearchOwnProps>(
  mapStateToProps,
  mapDispatchToProps
);

export const ProductsSearch = compose(withRouter, withExperienceEditorChromes, connectedToStore)(ProductsSearchComponent);
