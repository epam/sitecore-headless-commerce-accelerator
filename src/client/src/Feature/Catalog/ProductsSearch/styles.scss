//    Copyright 2020 EPAM Systems, Inc.
// 
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

@import 'Foundation/UI/common/scss/variables';

.products-search {
  position: relative;
  padding: 0 30px 50px 0;
  text-transform: uppercase;
  margin: 0;
  @media (max-width: $screen-sm-max) {
    text-transform: none;
    padding: 20px 0 0;
  }
  .search-wrap {
    position: relative;
    color: #666;
    &:before {
      font: normal normal normal 14px/1 FontAwesome;
      font-size: 20px;
      width: 20px;
      text-align: center;
      line-height: 1;
      content: "\f002";
      position: absolute;
      left: 15px;
      top: 50%;
      margin-top: -10px;
    }
  }
  form {
    position: relative;
    display: flex;
    margin-bottom: 0;
  }
  input[type=search] {
    margin: 0;
    -webkit-appearance: none;
    font-size: 14px;
    height: 43px;
    padding: 2px 55px 2px 36px;
    color: #666;
    border: 1px solid #e6e6e6;
    background: transparent none repeat scroll 0 0;
  }

  input[type=search]::-webkit-search-cancel-button {
    -webkit-appearance: none;
  }

  button {
    position: absolute;
    top: 50%;
    right: 0;
    height: 40px;
    width: 40px;
    margin: 0;
    padding: 0 15px;
    cursor: pointer;
    -webkit-transition: all .3s ease 0s;
    transition: all .3s ease 0s;
    -webkit-transform: translateY(-50%);
    transform: translateY(-50%);
    font-size: 16px;
    color: #666;
    border: #a1a5aa;
    border-left: 1px solid #a1a5aa61;
    -webkit-border-image: none;
    border-image: none;
    background: transparent none repeat scroll 0 0;
  }

  .resultLabels {
    display: flex;
    flex-direction: column;
    @media (min-width: $screen-sm-min) {
      flex-direction: row;
      justify-content: space-between;
    }
  }
}
  
