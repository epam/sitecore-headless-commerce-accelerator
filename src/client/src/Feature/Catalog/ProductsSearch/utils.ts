//    Copyright 2021 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import { tryParseUrlSearch } from 'Foundation/Base';

import { FACET_PARAMETER_NAME, KEYWORD_PARAMETER_NAME, Params } from 'services/productsSearch';

export const getKeywordFromSearch = (search: string) => {
  const parsedSearch = tryParseUrlSearch(search);

  return parsedSearch[KEYWORD_PARAMETER_NAME] || '';
};

export const buildQueryString = (value: string, params: Params) => {
  const newSearchQuery = [];

  newSearchQuery.push(`${KEYWORD_PARAMETER_NAME}=${value}`);

  if (params[FACET_PARAMETER_NAME]) {
    newSearchQuery.push(`${FACET_PARAMETER_NAME}=${encodeURIComponent(params[FACET_PARAMETER_NAME])}`);
  }

  return newSearchQuery.length !== 0 ? `?${newSearchQuery.join('&')}` : '';
};
