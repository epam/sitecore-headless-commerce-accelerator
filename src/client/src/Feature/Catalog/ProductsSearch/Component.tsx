//    Copyright 2020 EPAM Systems, Inc.
// 
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as React from 'react';

import * as Jss from 'Foundation/ReactJss';

import { FACET_PARAMETER_NAME, KEYWORD_PARAMETER_NAME } from 'Feature/Catalog/Integration/ProductsSearch';
import { tryParseUrlSearch } from 'Foundation/Base';

import { ProductsSearchOwnState, ProductsSearchProps } from './models';
import './styles.scss';

export default class ProductsSearch extends Jss.SafePureComponent<ProductsSearchProps, ProductsSearchOwnState> {
  public static getDerivedStateFromProps(props: ProductsSearchProps, state: ProductsSearchOwnState): ProductsSearchOwnState {
    const q = props.productSearchParams.q || '';
    if (state.initialKeyword !== q) {
      return {
        initialKeyword: q,
        keyword: q,
        submitted: true,
      };
    }

    return state;
  }

  constructor(props: ProductsSearchProps) {
    super(props);

    this.state = {
      initialKeyword: props.productSearchParams.q,
      keyword: this.getKeywordFromSearch(),
      submitted: false,
    };
  }

  protected safeRender() {
    const { keyword } = this.state;
    const { isLoading } = this.props;

    return (
      <section className="products-search">
      <h3 className="search-heading">Search</h3>
      <div className="search-wrap">
        <form onSubmit={(e) => this.handleFormSubmit(e)}>
          <input onChange={(e) => this.handleKeywordChange(e)} type="search" value={keyword} disabled={isLoading} placeholder="Search" />
          <button type="button" onClick={(e) => this.handleFormClear(e)}>
            <i className="fa fa-times" />
          </button>
        </form>
      </div>
      </section>
    );
  }

  private getKeywordFromSearch() {
    const search = this.props.history.location.search;
    const parsedSearch = tryParseUrlSearch(search);

    return parsedSearch[KEYWORD_PARAMETER_NAME] || '';
  }

  private handleKeywordChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      keyword: e.target.value,
      submitted: false,
    });
  }

  private handleFormSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();

    const { keyword, submitted } = this.state;

    if (submitted) {
      return;
    }

    this.startSearch(keyword);
  }

  private handleFormClear(e: React.MouseEvent<HTMLButtonElement>) {
    this.setState({ keyword: '', submitted: true });
    const { ChangeRoute, history } = this.props;
    const params = new URLSearchParams(history.location.search);
    params.delete('q');
    const encodeParams = encodeURIComponent(params.toString());
    const newUrl = `${history.location.pathname}?${encodeParams}`;
    ChangeRoute(newUrl);
  }

  private startSearch(keyword: string) {
    const params = this.props.productSearchParams;
    const newSearchQuery = [];
    if (keyword) {
      newSearchQuery.push(`${KEYWORD_PARAMETER_NAME}=${keyword}`);
    }
    if (params.f) {
      newSearchQuery.push(`${FACET_PARAMETER_NAME}=${encodeURIComponent(params.f)}`);
    }
    const newSearchQueryString = newSearchQuery.length !== 0 ? `?${newSearchQuery.join('&')}` : '';
    const { ChangeRoute, history } = this.props;
    this.setState({ submitted: true });
    ChangeRoute(`${history.location.pathname}${newSearchQueryString}`);
  }
}
