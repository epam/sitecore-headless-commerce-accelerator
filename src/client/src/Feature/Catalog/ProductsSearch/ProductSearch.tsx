//    Copyright 2021 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import React, { ChangeEvent, useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';

import { LoadingStatus } from 'Foundation/Integration';
import { ChangeRoute as changeRoute } from 'Foundation/ReactJss/SitecoreContext';

import { SidebarSearch } from 'components';
import { KEYWORD_PARAMETER_NAME, productSearchParams, productsSearchStatus } from 'services/productsSearch';

import { buildQueryString, getKeywordFromSearch } from './utils';

export const ProductsSearch = () => {
  const dispatch = useDispatch();

  const params = useSelector(productSearchParams);
  const status = useSelector(productsSearchStatus);

  const { search, pathname } = useLocation();

  const [value, setValue] = useState('');
  const [isSearchApplied, setIsSearchApplied] = useState(false);

  const loading = status === LoadingStatus.Loading;

  useEffect(() => {
    const queryValue = getKeywordFromSearch(search);

    if (!search) {
      dispatch(changeRoute(`${pathname}?${KEYWORD_PARAMETER_NAME}=`));
    }

    if (queryValue.length > 0) {
      setValue(queryValue);
      setIsSearchApplied(true);
    }
  }, [search]);

  const onChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
    setIsSearchApplied(false);
  }, []);

  const onClear = useCallback(() => {
    setValue('');
    setIsSearchApplied(true);

    const newSearchQueryString = buildQueryString('', params);

    dispatch(changeRoute(`${pathname}${newSearchQueryString}`));
  }, [isSearchApplied, pathname, params]);

  const onSubmit = useCallback(() => {
    if (!isSearchApplied) {
      const newSearchQueryString = buildQueryString(value, params);

      setIsSearchApplied(true);
      dispatch(changeRoute(`${pathname}${newSearchQueryString}`));
    }
  }, [dispatch, params, pathname, value]);

  return <SidebarSearch disabled={loading} value={value} onChange={onChange} onClear={onClear} onSubmit={onSubmit} />;
};
