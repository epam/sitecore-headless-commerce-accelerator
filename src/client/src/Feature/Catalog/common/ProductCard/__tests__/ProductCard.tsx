//    Copyright 2021 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { Product } from 'Feature/Catalog/Integration/ProductsSearch';

import { AddToCart, ColorVariants, Price, ProductCard, ProductName, ProductRating, ThumbnailSlider } from '../index';

const fallbackImageUrl = '/-/media/HCA/Project/HCA/no-image250x150.ashx';

const mockProduct: Product = {
  adjustedPrice: 240.17,
  brand: 'Spark Desktop PCs',
  currencySymbol: '$',
  customerAverageRating: null,
  description:
    'Helix’s Spark 6/1TB delivers you expanded storage, a quad-core processor, and the reliability of Helix computing at your fingertips. Finding an affordable tower with the performance you need is now here for your taking.',
  displayName: 'Spark Desktop—4GB Memory, 500GB Hard Drive',
  imageUrls: ['/-/media/Images/Habitat/6042236_01.ashx', '/-/media/Images/Habitat/6042236_02.ashx'],
  listPrice: 240.17,
  productId: '6042236',
  sitecoreId: 'bb515d29-4d7f-4699-1aa4-a4be54c69bad',
  stockStatusName: 'InStock',
  tags: ['desktop', '6gbram', '1tbhd', 'spark', 'tower'],
  variants: [
    {
      adjustedPrice: 240.17,
      brand: 'Spark Desktop PCs',
      currencySymbol: '$',
      customerAverageRating: null,
      description:
        'Helix’s Spark 6/1TB delivers you expanded storage, a quad-core processor, and the reliability of Helix computing at your fingertips. Finding an affordable tower with the performance you need is now here for your taking.',
      displayName: 'Spark Desktop—4GB Memory, 500GB Hard Drive',
      imageUrls: ['/-/media/Images/Habitat/6042236_01.ashx', '/-/media/Images/Habitat/6042236_02.ashx'],
      listPrice: 240.17,
      productId: '6042236',
      properties: { color: '', size: '' },
      stockStatusName: 'InStock',
      tags: ['desktop', '6gbram', '1tbhd', 'spark', 'tower'],
      variantId: '56042236',
    },
  ],
};

const productColors: Record<string, string> = {
  Beige: '#FFF8E7',
  Black: '#000000',
  Blue: '#3387d4',
  Camel: '#C19A6B',
  Charcoal: '#36454F',
  Clay: '#a76b29',
  Green: '#00FF00',
  Grey: '#808080',
  ['Light Blue']: '#ADD8E6',
  Mint: '#3EB489',
  Moss: '#ADDFAD',
  Orange: '#FF7F00',
  Pewter: '#8e9294',
  Pink: '#E34D67',
  Platinum: '#E5E4E2',
  Purple: '#9F00C5',
  Red: '#FF0000',
  Silver: '#C0C0C0',
  Spring: '#00FF7F',
  Stainless: '#E0DFDB',
  Teal: '#008080',
  Violet: '#7F00FF',
  White: '#F5F5F5',
  Yellow: '#FFFF00',
};

const mockStore = configureStore();
const store = mockStore({ shoppingCart: { data: {} } });

it('ProductCard renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <ProductCard fallbackImageUrl={fallbackImageUrl} product={mockProduct} productColors={productColors}>
          <ThumbnailSlider />
          <ColorVariants />
          <ProductName />
          <ProductRating />
          <Price />
          <AddToCart />
        </ProductCard>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
