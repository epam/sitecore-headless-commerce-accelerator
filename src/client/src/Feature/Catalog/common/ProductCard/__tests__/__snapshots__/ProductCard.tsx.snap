// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div
  className="ProductCard"
>
  <div
    className="ProductCard-ImageWrapper"
  >
    <div
      className="ProductCard-ImageInner"
    >
      <a
        href="/product/6042236"
        onClick={[Function]}
      >
        <div
          className="ImageSlider"
        >
          <div
            className="ImageSlider-SwiperContainer"
          >
            <div
              className="swiper-container"
            >
              <div
                className="swiper-wrapper"
              >
                <div
                  className="ImageSlider-Slide swiper-slide"
                >
                  <img
                    alt=""
                    className="ImageSlider-Image swiper-lazy"
                    src="/-/media/Images/Habitat/6042236_01.ashx"
                  />
                  <div
                    className="swiper-lazy-preloader swiper-lazy-preloader-white"
                  />
                </div>
                <div
                  className="ImageSlider-Slide swiper-slide"
                >
                  <img
                    alt=""
                    className="ImageSlider-Image swiper-lazy"
                    src="/-/media/Images/Habitat/6042236_02.ashx"
                  />
                  <div
                    className="swiper-lazy-preloader swiper-lazy-preloader-white"
                  />
                </div>
              </div>
              <div
                className="swiper-pagination"
              />
            </div>
          </div>
          <div
            className="ImageSlider-SwiperOverlay"
            onMouseLeave={[Function]}
          >
            <div
              className="ImageSlider-ButtonLeft"
              onMouseEnter={[Function]}
            />
            <div
              className="ImageSlider-ButtonRight"
              onMouseEnter={[Function]}
            />
          </div>
        </div>
      </a>
    </div>
  </div>
  <div
    className="ProductCard-ColorVariants"
  />
  <div
    className="ProductCard-ProductName"
  >
    <a
      href="/product/6042236"
      onClick={[Function]}
    >
      <span>
        Spark Desktopâ€”4GB Memory, 500GB Hard Drive
      </span>
    </a>
  </div>
  <div
    className="Rating Rating_theme_default Rating_size_m ProductCard-ProductRating"
  >
    <i
      className="IconFa fa StarIcon fa-star-o Rating-Star Rating-Star_active"
    />
    <i
      className="IconFa fa StarIcon fa-star-o Rating-Star Rating-Star_active"
    />
    <i
      className="IconFa fa StarIcon fa-star-o Rating-Star Rating-Star_active"
    />
    <i
      className="IconFa fa StarIcon fa-star-o Rating-Star Rating-Star_active"
    />
    <i
      className="IconFa fa StarIcon fa-star-o Rating-Star"
    />
  </div>
  <div
    className="ProductCard-Price"
  >
    <span
      className="ProductCard-FullPriceContainer"
    >
      <span
        className="ProductCard-Currency"
      >
        $
      </span>
      <span
        className="ProductCard-FullPrice"
      >
        240.17
      </span>
    </span>
  </div>
  <div
    className="ProductCard-AddToCart"
  >
    <button
      className="Button Button_buttonTheme_defaultSlide Button_buttonSize_m Button_fullWidth ProductCard-AddToCartButton"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      add to cart
    </button>
  </div>
</div>
`;
